src/api.ts

```diff
@@ -0,0 +1,637 @@
+    }));
+
+    // TODO: support requestParameters, authorizers, models and validators
+    this.method = props.resource.addMethod(props.httpMethod, integration, {
+      methodResponses,
+    });
+  }
+}
+
+/**
+ * An AWS API Gateway integration lets you integrate an API with an AWS service
+ * supported by Functionless. The request is transformed via VTL and sent to the
+ * service via API call, and the response is transformed via VTL and returned in
+ * the response.
+ *
+ * Only `application/json` is supported.
+ *
+ * TODO: provide example usage after api is stabilized
+ *
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-api-integration-types.html
+ */
+export class AwsApiIntegration<
+  Request extends ApiRequest,
+  MethodResponse,
+  IntegrationResponse
+> extends BaseApiIntegration {
+  private readonly request: FunctionDecl;
+  private readonly response: FunctionDecl;
+  private readonly errors: { [statusCode: number]: FunctionDecl };
+  readonly method;
+  public constructor(
+    readonly props: MethodProps,
+    /**
+     * Function that maps an API request to an integration request and calls an
+     * integration. This will be compiled to a VTL request mapping template and
+     * an API GW integration.
+     *
+     * At present the function body must be a single statement calling an integration
+     * with an object literal argument. E.g
+     *
+     * ```ts
+     *  (req) => fn({ id: req.body.id });
+     * ```
+     *
+     * The supported syntax will be expanded in the future.
+     */
+    request: (
+      $input: ApiGatewayInput<Request>,
+      $context: ApiGatewayContext
+    ) => IntegrationResponse,
+    /**
+     * Function that maps an integration response to a 200 method response. This
+     * is the happy path and is modeled explicitly so that the return type of the
+     * integration can be inferred. This will be compiled to a VTL template.
+     *
+     * At present the function body must be a single statement returning an object
+     * literal. The supported syntax will be expanded in the future.
+     */
+    response: (
+      $input: ApiGatewayInput<
+        { body: IntegrationResponse } & Omit<Request, "body">
+      >,
+      $context: ApiGatewayContext
+    ) => MethodResponse,
+    /**
+     * Map of status codes to a function defining the  response to return. This is used
+     * to configure the failure path method responses, for e.g. when an integration fails.
+     */
+    errors?: { [statusCode: number]: ($context: ApiGatewayContext) => any }
+  ) {
+    super();
+    this.request = validateFunctionDecl(request);
+    this.response = validateFunctionDecl(response);
+    this.errors = Object.fromEntries(
+      Object.entries(errors ?? {}).map(([k, v]) => [k, validateFunctionDecl(v)])
+    );
+
+    const role = new aws_iam.Role(props.resource, `Role_${props.httpMethod}`, {
+      assumedBy: new aws_iam.ServicePrincipal("apigateway.amazonaws.com"),
+    });
```

Comment:
Moved Role creation out of the integrations. This procedure runs for each integration, so it still results in one Role per method. Better to centralize logic for role creation than delegate to integrations - integrations should only worry about permissions.

This should be configurable. Excessive role creation may be undesirable.

Reply:
Thoughts on this implementation?
```ts
if (props.role) {
  // use the role passed in by the user.
  // TODO: should we try our best to ensure that the passed in Role has the correct assumedBy policy or leave it up to the user?
  //       -> if this is an IRole, we can't do anything
  //       -> if this is a Role, we can check the underlying CfnResource's policy and update it
  return props.role;
}
// by default, create a Role for each Resource's HTTP Method
const roleResourceName = `Role_${props.httpMethod}`;
// the Method's Role is stored as a singleton on the HTTP Resource using the naming convention, `Role_<method>`, e.g. `Role_GET`.
const roleSingleton = props.resource.node.tryFindChild(roleResourceName);
if (roleSingleton) {
  return roleSingleton as aws_iam.Role;
} else {
  return new aws_iam.Role(props.resource, roleResourceName, {
    assumedBy: new aws_iam.ServicePrincipal("apigateway.amazonaws.com"),
  });
}
```

Reply:
there is a `singletonConstruct` method that should work for you, otherwise, yeah, looks good.